### DISCUSSION TITLE
Does AI tagging look at and use the already created tags?

### DISCUSSION THREAD

**Discussion by kriskras69** [karakeep-app/karakeep#702] [Q&A] [answered]
I don't use AI tagging but want to try it out, hence my question.
Would it be possible to use 'AI Tagging rules' to tell Hoarder to look at my current tags when
auto-tagging?

**Comment by MohamedBassem** [comment, marked as answer]
We have this in the nightly build and it'll be available in the next release.

**Reply by kamtschatka** [reply]
see: https://github.com/hoarder-app/hoarder/pull/612 and https://github.com/hoarder-app/hoarder/issues/111

**Reply by kriskras69** [reply]
Wauw, thats good news.
May I take this opportunity to thank you, and everyone involved, for this great app? I'm sure I may so I do. 👍

**Comment by taylorbanks** [comment]
how is this enabled? I find that even on 0.26.0, every new entry seems to just create new tags. there doesn't seem to be any logic to prefer pre-existing tags (or avoid creating similar tags when relevant tags already exist)

**Reply by kriskras69** [reply]
You can append additional instructions to the prompt used for automatic tagging, in the AI Settings (in the User Settings screen)
You can use the placeholders $tags, $aiTags, $userTags in the prompt. These placeholders will be replaced with all tags, ai generated tags or human created tags when automatic tagging is performed (e.g. [karakeep, computer, ai])

**Reply by ben-kenney** [reply]
Is this how it should be done?

in my AI settings I’ve added the following tagging rule:

> - Try to reuse existing tags if applicable. The existing tags are $userTags

The thing is I don’t know how to verify that this is actually working. It would be nice to see the list of tags that this actually generates.

**Reply by kriskras69** [reply]
I personal have multiple instances of Karakeep.  I dont want to break anything on my main so i try stuff on a different Karakeep stack.

**Comment by ryanpeach** [comment]
I find that on 0.31.0 https://github.com/karakeep-app/karakeep/pull/2444 all we added was the ability to force the AI to pick from curated tags. But when curated tags aren't set, I still want the prompt to be able to see the available tags and _prefer_ to pick from those already available. This might need some fancy prompt engineering but it's not quite what was offered in the PR.

**Comment by ryanpeach** [comment]
Adding "Prefer tags from $tags" to my Tagging Rules was enough. That's basically what I wanted.
